### PR TITLE
Fixes #32572 - PBKDF2 password hashing support

### PR DESCRIPTION
**Pull request by lzap**
I changed password and personal token hashing from very weak SHA1 to quite strong bcrypt. But I haven't realized bcrypt is based on blowfish whith is not approved by FIPS/NIST standards. Therefore this patch adds PBKDF2 password hashing while keeping ability to switch between bcrypt and PBKDF2 depending on the use case. Bcrypt is still considered very strong, however, being not approved by FIPS this cannot be used by some of our customers.

I am filing this as a draft, I am currently talking to our crypto team to confirm which hash algorithm to use, how many cycles/passes should be set. BCrypt library has a nice feature to calibrate for a specific time, however this is not available in the PBKDF2 OpenSSL implementation. We can start some decent value like 4096 and probably make this a configurable option.

Recommended setting by the crypto team: PBKDF2-HMAC-SHA512, 256 bit salt and hash minimum and iteration set by performing benchmark to a reasonable count (thousands or tens of thousands). Here are my test results:

https://lukas.zapletalovi.com/2021/05/finding-the-right-cost-for-bcryptpbkdf2.html

https://github.com/theforeman/foreman/pull/5633

Reviewer: To test this, change the password to PBKDF2 and then change password for a user, ensure that in database the password in encrypted in the new format `$pbkdf2$.....$` and that it actually works (user can log in).

### CODE REVIEW

File: test/models/user_test.rb [hunk 1]
```diff
@@ -60,7 +60,7 @@ def setup
   should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html, :punctuation, :cyrillic, :utf8]).values).for(:lastname)
   should allow_value('A$+#APRocky').for(:login)
   should allow_value(*valid_name_list).for(:description)
-  should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html])).for(:password)
+  should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html]).values).for(:password)
```
Comment (lzap) [inline test/models/user_test.rb:63]: While I was testing this patch, OpenSSL library was erroring out because this code actually tries to store the following array into password:

```
UserTest#test_: User should allow :password to be ‹[:alpha, "NgdMvTEoLiXOWjwlyNYACICCAmahfTJQJHQuYSEPIFSmtWVmP"]›, ‹[:alphanumeric, "sLntjP7kT48Hn7bZ8UPZ5mFQRn8L1GnsHbuY3"]›, ‹[:cjk, "鄢堆阸牬黋垹菍漎凇詹攙啝將貎肏"]›, ‹[:cyrillic, "ӝѧИѷ҅ЁКЙЬЦЦӦЕӊҺэЖ҇аѣҙѣТр҂ѲгӼКЅҧЩѦдҜӋөѨ"]›, ‹[:latin1, "ÉÇþÃíúÕüÑõÖÇëÅîÉøôáòÛèôËÈöò"]›, ‹[:numeric, "150737650672785103224408701503244304007721688444"]›, ‹[:utf8, "𦟇𡫹𨋚𥧎𦊀繐꾆"]›, or ‹[:punctuation, "(}"]›. :
```

I think all these RFauxFactory tests in this file, and in the whole codebase, are incorrect:

```
test/models/hostgroup_test.rb
12:    *RFauxFactory.gen_strings(1..245, exclude: [:html, :punctuation]).values,

test/models/host_test.rb
8:    *RFauxFactory.gen_strings(1..255, exclude: [:html, :utf8]).values,

test/models/organization_test.rb
15:      *RFauxFactory.gen_strings(1..242, exclude: [:html]).values,

test/models/location_test.rb
11:      *RFauxFactory.gen_strings(1..246, exclude: [:html]).values,

test/models/user_test.rb
58:  should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html, :punctuation, :cyrillic, :utf8]).values).for(:firstname)
60:  should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html, :punctuation, :cyrillic, :utf8]).values).for(:lastname)
63:  should allow_value(*RFauxFactory.gen_strings(1..50, exclude: [:html]).values).for(:password)
65:  should_not allow_value(*RFauxFactory.gen_strings(51)).for(:firstname)
67:  should_not allow_value(*RFauxFactory.gen_strings(51)).for(:lastname)

test/test_helper.rb
50:    *RFauxFactory.gen_strings(1..255, exclude: [:html]).values.map { |x| x.truncate_bytes(255, omission: '') },
```

Shall I just fix these or file a different PR? This line must be changed tho, OpenSSL will fail when array is attempted to be encrypted instead of string.

Comment (ares) [reply test/models/user_test.rb:63]: I'd be ok with fixing this in here

**Comment by ares** [comment]
Hello @lzap, given this is a draft for a long time we'd like to know if you have a plan to move this out of the draft in a foreseeable future.

If we don't get a reply in 2 weeks, we'll close the PR but it can always be reopened.

**Comment by tbrisker** [comment]
2 weeks have passed, closing. Feel free to reopen when you get back to this.

File: lib/foreman/password_hash.rb [hunk 2]
```diff
@@ -9,12 +9,32 @@
 #
 module Foreman
   class PasswordHash
+    class PBKDF2Implementation
+      SIZE = 39
+      ITERATIONS = 4096
```
Comment (lzap) [inline lib/foreman/password_hash.rb]: Cost should be also configurable, like for bcrypt. The safe default value should be probably 100,000 iterations.

**Comment by lzap** [comment]
So I will work on this tomorrow, will finish this, implement new settings but I will keep PBKDF2-HMAC-SHA1 as the default implementation for now, iterations will be configurable tho.

**Comment by lzap** [comment]
Rebased.

File: app/registries/foreman/settings/auth.rb [hunk 3]
```diff
@@ -118,11 +124,22 @@
       description: N_("Log out idle users after a certain number of minutes"),
       default: 60,
       full_name: N_('Idle timeout'))
+    setting('password_hash',
+      type: :string,
+      description: N_("Password hashing algorithm. A password change is needed effect existing passwords."),
+      default: "pbkdf2sha1",
+      full_name: N_('Password hashing algorithm'),
+      collection: proc { HASHING_ALGORITHMS })
```
Comment (ezr-ondrej) [inline app/registries/foreman/settings/auth.rb]: Can't this setting be in the file? I think that we want the users to run some job that will transform the passwords in DB after changing this setting, right? so it should require a restart anyway, so having it in the UI might be harder to understand that user need to run something after changing the setting.

File: test/fixtures/settings.yml [hunk 4]
```diff
@@ -418,3 +418,13 @@ attribute98:
...
+attribute99:
+  name: password_hash
+  category: Setting::Auth
+  default: 'pbkdf2sha1'
+  description: "Password hash"
+attribute100:
+  name: pbkdf2_cost
+  category: Setting::Auth
+  default: 1000
+  description: "PBKDF2 cost"
```
Comment (ezr-ondrej) [inline test/fixtures/settings.yml]: fixutures for settings are not needed anymore :)

File: app/registries/foreman/settings/auth.rb (hunk 3 above)
Comment (lzap) [reply app/registries/foreman/settings/auth.rb]: No action other than setting this value is needed, you cannot do any migration (or anything pretty much) as you don't have the actual passwords which is actually why you hash passwords in the first place. Only users can "migrate" themselves by re-entering their passwords again into the new algorithm.

Note after the setting change, old passwords hashed with old values will continue to work, that is how this is designed and how it should be - the same behavior is for example in Linux (PAM).

To solve this, systems typically have a "request password change on next login" feature which is out of scope for this patch - I don't have capacity to implement such feature. Thus I went for the approach of describing this in the setting - I am literally asking the admin to ask for password changes.

File: test/fixtures/settings.yml (hunk 4 above)
Comment (lzap) [reply test/fixtures/settings.yml]: Like not at all? Why the file is still present?

Comment (lzap) [reply test/fixtures/settings.yml]: Removed.

**Comment by lzap** [comment]
Rebased, fixed tests.

File: app/registries/foreman/settings/auth.rb [hunk 5]
```diff
@@ -1,3 +1,9 @@
+HASHING_ALGORITHMS = {
+  "sha1" => _("SHA1"),
+  "bcrypt" => _("BCrypt"),
+  "pbkdf2sha1" => _("PBKDF2 SHA1"),
+}
+
```
Comment (ezr-ondrej) [inline app/registries/foreman/settings/auth.rb]: This is incorrect, the translations will be evaluated on code load.
Translated strings should be inside of the proc, that is why we have it in the first place.
Why are we translating the strings anyway? and could we not just define this inplace instead of having global constant?

**Review by ezr-ondrej — COMMENTED** [review]
I have just an issue with the global constant, but apart of that looks good to me :+1:

File: app/registries/foreman/settings/auth.rb (hunk 3 above)
Comment (ezr-ondrej) [reply app/registries/foreman/settings/auth.rb]: Right, make sense, thanks for the explanation! :+1:

File: test/unit/tasks/seeds_test.rb [hunk 6]
```diff
@@ -15,6 +15,7 @@ class SeedsTest < ActiveSupport::TestCase
     Setting.stubs(:[]).with(:bcrypt_cost).returns(1)
     Setting.stubs(:[]=).with(:bcrypt_cost, anything).returns(true)
```
Comment (ezr-ondrej) [inline test/unit/tasks/seeds_test.rb:19]: we will probably want to stub the new settings as well, right?

**Comment by ezr-ondrej** [comment]
@lzap you're changing the default here, can we add an release/upgrade note for that?

File: test/unit/tasks/seeds_test.rb (hunk 6 above)
Comment (lzap) [reply test/unit/tasks/seeds_test.rb:19]: This is only for seed test, no need to stub unused settings I guess.

Comment (lzap) [reply test/unit/tasks/seeds_test.rb:19]: Whatever I can add them all just to be sure to prevent issues.

**Comment by lzap** [comment]
Rebased, I set the default hash back to bcrypt.

File: package.json [hunk 7]
```diff
@@ -28,7 +28,7 @@
     "react-intl": "^2.8.0"
   },
   "devDependencies": {
-    "@babel/core": "^7.7.0",
+    "@babel/core": "^7.17.8",
```
Comment (evgeni) [inline package.json]: this doesn't seem related (and earned you a special packaging review, which shouldn't be needed otherwise?)

Comment (lzap) [reply package.json]: Thanks, fixed.

File: app/services/foreman/password_hash.rb [hunk 8]
```diff
@@ -40,18 +59,25 @@ def hash_secret(password, salt)
...
       else
         raise(Foreman::Exception.new(N_("Unknown password hash method: %s"), implementation))
       end
     end
 
     def self.detect_implementation(password_string)
-      password_string.start_with?('$2') ? :bcrypt : :sha1
+      return :pbkdf2sha1 if password_string.start_with?('$pbkdf2sha1')
+      return :bcrypt if password_string.start_with?('$2')
+      :sha1
```
Comment (ekohl) [inline app/services/foreman/password_hash.rb]: Personally I think this is more readable
```suggestion
      if password_string.start_with?('$pbkdf2sha1')
        :pbkdf2sha1 
      elsif password_string.start_with?('$2')
        :bcrypt
      else
        :sha1
      end
```

File: app/services/foreman/password_hash.rb [hunk 9]
```diff
@@ -9,12 +9,31 @@
 #
 module Foreman
   class PasswordHash
+    class PBKDF2Implementation
+      SIZE = 39
```
Comment (ekohl) [inline app/services/foreman/password_hash.rb]: Can you comment on what this size is? I see `generate_salt` accepts a size parameter but `hash_secret` doesn't.

File: app/services/foreman/password_hash.rb (hunk 8 above)
Comment (lzap) [reply app/services/foreman/password_hash.rb]: Sure.

File: app/services/foreman/password_hash.rb (hunk 9 above)
Comment (lzap) [reply app/services/foreman/password_hash.rb]: Renamed to much more reasonable `DEFAULT_SALT_LENGTH`.

**Comment by lzap** [comment]
`NameError: uninitialized constant ActiveRecord::Base::EnsureNotUsedBy`

De fak?!

**Comment by lzap** [comment]
Fixed also seed tests, rebased.

**Review by ezr-ondrej — APPROVED** [review]
Thanks @lzap ! :+1:

**Comment by lzap** [comment]
https://github.com/theforeman/foreman-documentation/issues/1194
